drivers: display: Add 3wire mode support in st7735 controller

ozersa opened the pull request:
st7735 display controller can driver display both over 3wire or 4wire mode.
Existing st7725r.c display controller only supports 4wire mode. This commit add 3wire mode support in the controller.

Both 3wire and 4wire mode tested with [MAX32672EVKIT ](https://www.analog.com/en/resources/evaluation-hardware-and-software/evaluation-boards-kits/max32672evkit.html) which has [CFAF128128B1-0145T](https://www.crystalfontz.com/product/cfaf128128b10145t-128x128-graphic-tft-spi) display on board.

- 3-wire, 9-bit SPI mode. The D/C pin is not used. The 9th SPI bit selects the data or command register.
- 4-wire, 8-bit SPI mode. The D/C pin is used to select the data or command register.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/46590392/0079372f-e439-4319-ad66-c3610425bdf5)

ozersa committed:
style: drivers: display: Run clang-format for st7735r driver

display_st7735r.c/.h files are reformatted with clang-format to
clearly demonstrate additions on next commits.

There is no any additional update.

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa committed:
drivers: display: Update st7735 to work with MIPI DBI

This commit update st7735 display driver to it works
with MIPI driver

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

MaureenHelm commented on drivers/display/display_st7735r.c:
```diff
@@ -67,32 +67,37 @@ static void st7735r_set_lcd_margins(const struct device *dev, uint16_t x_offset,
+		tx_buf.len = tx_count;
+
+		/* Drive Data/Comman pin  */
+		gpio_pin_set_dt(&config->cmd_data, (is_data ? 0 : 1));
```
> Some additional tab added in some lines to make next commit more clear.

Don't do this. Each commit should be able to pass code style checks

ozersa replied:
Thanks Maureen for feedback.
Sure.

MaureenHelm commented on dts/bindings/display/sitronix,st7735r.yaml:
```diff
@@ -131,3 +132,14 @@ properties:
       This option is convenient for supporting displays with bugs
       where the actual color is different from the pixel format of MADCTL.
+
+  spi-mode-3wire:
+    type: boolean
+    description: |
+      Enable 3 wire control mode
+
+      3-wire, 9-bit SPI mode. The D/C pin is not used.
+      The 9th SPI bit selects the data or command register.
+
+      4-wire, 8-bit SPI mode.
+      The D/C pin is used to select the data or command register.
```
This property should be moved to `dts/bindings/spi/spi-device.yaml` and renamed to `spi-3wire` to be consistent with the Linux kernel.

ozersa replied:
Today I see [st7789v](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/display/sitronix%2Cst7789v.yaml#L21) driver, in this driver this option implicitly handled by "gpio command". 
And MIPI DIB API also does not need additional parameter to select 3/4 wire mode.  Please see: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/mipi_dbi/Kconfig.spi#L16

So it might be better if we totally remove this additional parameter.

Anyway if you think it will be better  please let me know.

danieldegrasse reviewed (changes requested):
This can be implemented using the MIPI DBI API, which already implements support for SPI 3 wire mode: https://docs.zephyrproject.org/latest/hardware/peripherals/mipi_dbi.html.

Instead of implementing 3 wire support in the driver, can you instead update this driver to use the MIPI DBI API? That should enable 3 wire support to be added relatively easily

Just a note- although the code the 3 wire support is present in the MIPI DBI SPI mode driver, I don't have a 3 wire mode display to test support with. If you run into issues with the implementation please let me know, I'd be happy to help get this moved over

ozersa commented:
> This can be implemented using the MIPI DBI API, which already implements support for SPI 3 wire mode: https://docs.zephyrproject.org/latest/hardware/peripherals/mipi_dbi.html.
> 
> Instead of implementing 3 wire support in the driver, can you instead update this driver to use the MIPI DBI API? That should enable 3 wire support to be added relatively easily
> 
> Just a note- although the code the 3 wire support is present in the MIPI DBI SPI mode driver, I don't have a 3 wire mode display to test support with. If you run into issues with the implementation please let me know, I'd be happy to help get this moved over

Thanks for the feedback, I just get a chance to revisit this
It seems  MIPI DBI driver requires some updates, I have update it to work with MIPI API but st7735r.c driver includes some spi_release_dt(...) call which not easily be replaced with MIPI driver update. I have not found a appropriate way for them yet.

Please see my updated here: https://github.com/zephyrproject-rtos/zephyr/commit/2e754db57cf38945c91df238f9e5bff3c6861be5

Also today I see [st7789v](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_st7789v.c) driver which handle 3/4 wire mode in the driver, please see: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_st7789v.c#L71

Does it make sense we follow st7789v.c format and then in future convert them to MIPI DBI API?

danieldegrasse commented:
> Thanks for the feedback, I just get a chance to revisit this
It seems MIPI DBI driver requires some updates, I have update it to work with MIPI API but st7735r.c driver includes some spi_release_dt(...) call which not easily be replaced with MIPI driver update. I have not found a appropriate way for them yet.

I'll be honest- this is exactly why I want to convert drivers like this. I unfortunately don't have a 3 wire display to test with, but we need to prove out support for the MIPI DBI 3 wire mode. If we need to make changes to the MIPI DBI driver implementation to get things working with 3 wire mode, PRs like this will catch that. If I converted this driver to use the MIPI DBI API, would you be willing to help test 3 wire support with your display?

ozersa committed:
drivers: mipi: SPI len shall 2 for 3wire mode

3wire mode requires 9bits, and if data being
bigger than 8bits len shall be 2.

Signed-off-by: Sadik Ozer <sadik.ozer@analog.com>

ozersa commented:
> > Thanks for the feedback, I just get a chance to revisit this
> > It seems MIPI DBI driver requires some updates, I have update it to work with MIPI API but st7735r.c driver includes some spi_release_dt(...) call which not easily be replaced with MIPI driver update. I have not found a appropriate way for them yet.
> 
> I'll be honest- this is exactly why I want to convert drivers like this. I unfortunately don't have a 3 wire display to test with, but we need to prove out support for the MIPI DBI 3 wire mode. If we need to make changes to the MIPI DBI driver implementation to get things working with 3 wire mode, PRs like this will catch that. If I converted this driver to use the MIPI DBI API, would you be willing to help test 3 wire support with your display?

As I mentioned, I have already converted driver to mipi dbi after your first comment, it works with 8bit mode (4wire). but not work for 3wrire mode, seems stop clocking after some bytes, I do not know how should  handle spi_release_dt(...) function which exist in old version of st7735r.c driver. I  believe this is the reason of the failure for 3wire mode.  I am going to check SPI line with logic analyzer.
Now this PR  includes MIPI DIB conversion, please take a look, and if you are not happy with the conversion to MIPI DBI driver i am fine, please feel free to convert it, incase of need I can apply test on my board.

danieldegrasse commented:
> As I mentioned, I have already converted driver to mipi dbi after your first comment, it works with 8bit mode (4wire). but not work for 3wrire mode, seems stop clocking after some bytes, I do not know how should handle spi_release_dt(...) function which exist in old version of st7735r.c driver. I believe this is the reason of the failure for 3wire mode. I am going to check SPI line with logic analyzer.
Now this PR includes MIPI DIB conversion, please take a look, and if you are not happy with the conversion to MIPI DBI driver i am fine, please feel free to convert it, incase of need I can apply test on my board.

Thank you- I really appreciate you looking at this. I took a look at converting this display as well: https://github.com/zephyrproject-rtos/zephyr/pull/70583. I've added the ability to specify the mode in devicetree with that PR- perhaps you could try using 3 wire mode with that change? I added some code to handle `spi_release_dt`, but I unfortunately can't test it as I don't have this display.

ozersa commented:
Thanks for the update, today I tried to test it, I cherry picked your commits from nxp fork, but encounter confliction between multiple zephyr forks,
We have an internal fork (it is behind the upstream), you are working on nxp fork, So that it will be better for me check it after we fetch upstream in next weeks.

ozersa commented:
Superseded by https://github.com/zephyrproject-rtos/zephyr/pull/70583
